### DISCUSSION TITLE
options() in an extension package

### DISCUSSION THREAD

**Discussion by EvaMaeRey** [ggplot2-extenders/ggplot-extension-club#48] [Q&A]
What are best practices for using options or similar.

**Comment by teunbrand** [comment]
I don't really like using the options. It feels like some global state that departs from the grammar.
That said, probably use `getOption()` explicitly in the function arguments, so that users can see that in the usage section.
Example:

```r
my_function <- function(x, opt = getOption("mypackage.opt", default = NULL)) {
  opt <- opt %||% compute_default(x)
  ...
}
```

**Reply by yjunechoe** [reply]
I've never noticed this but that's a nice design! Ex:

```r
> ggplot2::scale_fill_discrete
function (..., type = getOption("ggplot2.discrete.fill")) 
{
    type <- type %||% scale_fill_hue
    <...>
}
```

**Comment by yjunechoe** [comment]
I think the answer also depends heavily on whether the package plans to simply *read* ("get") options, or also *write* ("set") options to different values in different contexts (aside from initializing them).

I keep bookmarked some advice for taming the side-effect behavior of `options()` in:
- R Packages: https://r-pkgs.org/code.html#sec-code-r-landscape
- Tidy design principles: https://design.tidyverse.org/side-effects.html

Edit: for the record, I also don't really like using `options()` other than for trivial behaviors that specifically target the user *experience* (vs. change a package's feature) like the verbosity of messages/warnings.

**Comment by EvaMaeRey** [comment]
One place where I felt really grateful for options, and where it also felt like it made sense and was justified is `options(lubridate.week.start = 1)` so that wday will switch from Sunday to Monday.  I think a really global switch needs to be possible for this kind of thing.

That said, my use case might not be as compelling.  Something like...

options(ggregion.geo.ref.data = northcarolina_counties, ggregion.required.aes = "fips|county.name"

Then exported functions geom_region() and friends would be able to just work in the northcarolina county context, I think.

**Reply by EvaMaeRey** [reply]
I use getOption and options, and find it pretty nice.  So then you can use the same syntax (geom_region()) to make plots about North Carolina, Chile, the world, Netherlands, etc (I do have one problem case usamaps - not sure what is going on there...)

https://github.com/EvaMaeRey/sf2stat?tab=readme-ov-file#wrapping-up-more

``` r
options(sf2stat.ref_data = nc_ref,
        sf2stat.required_aes = "fips|county_name")

# north carolina counties....
read.csv("nc-midterms.csv") |>
  ggplot() + 
  aes(county_name = str_to_title(desc_county)) + 
  stamp_region(fill = 'darkgrey') + 
  geom_region() + 
  aes(fill = n/1000) + 
  stamp_region_outline(
    keep_id = "Mecklenburg",
    color = "orange",
    linewidth = 1) + 
  geom_region_text(check_overlap = T,
                   color = "whitesmoke")
```

![image](https://github.com/user-attachments/assets/c6c6845f-aa43-4e21-a267-ee3e2fa69a06)


``` r
## set options: reference data and required aes
chilemapas::generar_regiones() %>% 
  mutate(numero_region = as.numeric(codigo_region)) %>% 
  options(sf2stat.ref_data = .,
          sf2stat.required_aes = "codigo_region|numero_region")

## plot
chilemapas::censo_2017_comunas %>% 
  mutate(region = str_extract(codigo_comuna, "..")) %>% 
  summarise(pop = sum(poblacion), .by = c(region, sexo)) %>% 
  ggplot() + 
  aes(codigo_region = region, fill = pop/100000) +
  geom_region(linewidth = .01) + 
  facet_wrap(~sexo) + 
  scale_fill_viridis_b(transform = "log") + 
  stamp_region_outline(color = "red", 
                       keep_id = "05")

```

![image](https://github.com/user-attachments/assets/69b83586-6afb-4cbb-9cc1-733ff05eaa26)


``` r

library(tmap)

data(World)

## set options: reference data and required aes
World %>% 
  select(country_name = name,
         iso3c = iso_a3) %>% 
  options(sf2stat.ref_data = .,
          sf2stat.required_aes = "country_name|iso3c")
  
# plot
tidyr::world_bank_pop %>% 
  filter(indicator == "SP.POP.GROW") %>% 
  ggplot() + 
  aes(iso3c = country) + 
  geom_region() + 
  aes(fill = `2000`) + 
  scale_fill_viridis_c() + 
  labs(title = "Population Growth, 2000")

```

![image](https://github.com/user-attachments/assets/d0e06539-5e89-4c11-bf9b-9ac5adec5552)

``` r
data("NLD_prov")

## set options: reference data and required aes
NLD_prov %>% 
  select(prov_name = name, prov_code = code) %>% 
  options(sf2stat.ref_data = ., 
          sf2stat.required_aes = "prov_code|prov_name")

# plot
NLD_prov %>% 
  sf::st_drop_geometry() %>% 
  ggplot() + 
  aes(prov_code = code) + 
  geom_region() + 
  aes(fill = population/100000) + 
  geom_region_text(check_overlap = T,
                   size = 2, 
                   color = "whitesmoke") + 
  aes(label = round(population/100000, 3))

```

![image](https://github.com/user-attachments/assets/f350e124-ff18-4183-8740-62822413530b)

**Reply by EvaMaeRey** [reply]
But might not feel as easy/intuitive to use as

my_world_data_countries |>
ggplot() + 
aes(country_name = country) + 
geom_country()

**Reply by EvaMaeRey** [reply]
wrapping up option setting makes things feel pretty clean (I still need to consider warnings away from the options usa at all). 
 So below, geom_region will treat countries as regions, and the country geometries come from the rnaturalearth package...

![image](https://github.com/user-attachments/assets/ca2caa04-86b3-4e9f-b6e0-38bd7bf11260)

**Reply by teunbrand** [reply]
Out of curiosity: where did France go? I think I saw another map recently that also had France missing.

**Reply by EvaMaeRey** [reply]
Hmm.  My default is to blame surprising missingness on country names joins, but join is on iso code which should perform nicely.  So 1) codes may not be aligned or data is missing for france for that year and indicator in tidyr::worldbank. 

I see that France does appear on this internet usage map - where aes is country_name, so suspect rnaturalearth to be behaiving. 

![image](https://github.com/user-attachments/assets/3ac898d9-fd75-4644-add0-5711dde354d9)

stamp_region() is used in the internet usage chart to fill in unfound countries, but I guess I wasn't bothered enough by france's not being matched up.

**Reply by EvaMaeRey** [reply]
in the internet plot, geom_region does an inner join with geo reference data.  The dark grey is where country is found in the data has no fill value (internet or population is missing) and the very light grey showing through for stamp_region where no region match is found in the plot data, but region exists in reference data. Maybe I should switch stamp fill default to sf's NA fill default.

**Reply by EvaMaeRey** [reply]
Nope... rnaturalearth is the culprit. 

![image](https://github.com/user-attachments/assets/2485914b-6386-4f58-861f-0346839c5db9)

**Reply by EvaMaeRey** [reply]
set_region_country_rnaturalearth <- function(scale = "small"){

rnaturalearth::ne_countries(
  scale = scale, returnclass = "sf") |> 
  select(country_name = sovereignt, iso3c = iso_a3) |>
  mutate(country_name = ifelse(country_name == "United States of America", "United States", country_name)) |>
  mutate(iso3c = ifelse(country_name == "France", "FRA", iso3c)) |>
  options(sf2stat.ref_data = _,
          sf2stat.required_aes = "country_name|iso3c") 
  
}

![image](https://github.com/user-attachments/assets/e536549f-39c0-4c1f-8e33-d2b94382cdf2)

**Reply by EvaMaeRey** [reply]
Annotation layer feature...

![image](https://github.com/user-attachments/assets/121aa0b9-ab09-4f3b-b8d8-f9e45d64254a)

**Reply by EvaMaeRey** [reply]
Similar 'stamp' code below, but with fill scale informing about regions.

![image](https://github.com/user-attachments/assets/60f9fec9-0a57-4877-8d5f-2c42e6c5bb55)

**Reply by EvaMaeRey** [reply]
I like the possibility of  using this options strategy in modeling context too. Where you might want to define the model of interest, for example y~x+cat 😺 and lm, and then plot feature incrementally, model fit across x, fitted values, residuals, etc. where each is a layer.
